packages/roosterjs-editor-dom/lib/utils/setColor.ts

```diff
@@ -34,5 +51,72 @@ export default function setColor(
                 element.dataset[dataSetName] = modeIndependentColor.lightModeColor;
             }
         }
+
+        if (isBackgroundColor && shouldAdaptTheFontColor) {
+            adaptFontColorToBackgroundColor(element, isDarkMode);
+        }
+    }
+}
+
+/**
+ * Change the font color to white or some other color, so the text can be visible with a darker background
+ * @param element The element that contains text.
+ */
+function adaptFontColorToBackgroundColor(element: HTMLElement, isDarkMode?: boolean) {
+    if (element.firstElementChild?.hasAttribute('style')) {
+        return;
+    }
+    const backgroundColor = element.style.getPropertyValue('background-color');
+
+    if (!backgroundColor) {
+        return;
+    }
+
+    const isADarkOrBrightOrNone = isADarkOrBrightColor(backgroundColor);
+    const fontColorInDarkMode = element.dataset[DarkModeDatasetNames.OriginalStyleBackgroundColor];
+    switch (isADarkOrBrightOrNone) {
+        case ColorTones.DARK:
+            element.style.color = WHITE;
+            break;
+        case ColorTones.BRIGHT:
+            //Save this value so transformColor can adjust the color when switch from dark to light mode
+            element.dataset[DarkModeDatasetNames.OriginalStyleColor] = WHITE;
```

Comment:
I believe you should either set this property for both case ColorTones.DARK and here, or neither of them.

And you can read the background color like:
const lightModeBackgroundColor = isDarkMode ? element.dataset[DarkModeDatasetNames.OriginalStyleBackgroundColor] || element.dataset[DarkModeDatasetNames.OriginalAttributeBackgroundColor] : backgroundColor;

Now you have the back color always for light mode, then calculate color tone from it, and set the text color to the cell, and delete element.dataset[DarkModeDatasetNames.OriginalStyleColor] and element.dataset[DarkModeDatasetNames.OriginalAttributeColor].

After that, the setShadeColor code can call dark mode transform code to transform to dark mode if necessary.